### PR TITLE
Feature/multi token support

### PR DESCRIPTION
**Pull request by BrianCorbin**
Soundtrack of this PR: [link to song that really fits the mood of this PR]()

### Motivation

This adds support for multiple tokens in Full Service simultaneously

### In this PR
* updated API where relevant to support token ids
* updated service and db layer to support multi tokens

### CODE REVIEW

File: full-service/src/db/transaction_log.rs [hunk 1]
```diff
@@ -350,11 +351,11 @@ impl TransactionLogModel for TransactionLog {
         // Verify that the TxProposal is well-formed according to our
         // assumptions about how to store the sent data in our wallet
         // (num_output_TXOs = num_outlays + change_TXO).
-        if tx_proposal.tx.prefix.outputs.len() - tx_proposal.outlays.len() > 1 {
-            return Err(WalletDbError::UnexpectedNumberOfChangeOutputs);
-        }
+        // if tx_proposal.tx.prefix.outputs.len() - tx_proposal.outlays.len() > 1 {
```
Comment (Shramp) [inline full-service/src/db/transaction_log.rs:354]: delete commented code

Comment (sugargoat) [inline full-service/src/db/transaction_log.rs:354]: Should this be commented out?

File: full-service/src/db/transaction_log.rs [hunk 2]
```diff
@@ -374,9 +375,10 @@ impl TransactionLogModel for TransactionLog {
             .values(&new_transaction_log)
             .execute(conn)?;
 
-        for utxo in tx_proposal.utxos.iter() {
-            let txo_id = TxoID::from(&utxo.tx_out);
-            Txo::update_key_image(&txo_id.to_string(), &utxo.key_image, None, conn)?;
+        for txo in tx_proposal.input_txos.iter() {
+            let txo_id = TxoID::from(&txo.tx_out);
+            Txo::update_key_image(&txo_id.to_string(), &txo.key_image, None, conn)?;
+            // Txo::update_key_image(&txo_id.to_string(), &utxo.key_image, None, conn)?;
```
Comment (Shramp) [inline full-service/src/db/transaction_log.rs:381]: delete commented code

File: full-service/src/db/transaction_log.rs [hunk 3]
```diff
@@ -405,12 +410,12 @@ impl TransactionLogModel for TransactionLog {
...
-        // Verify that the TxProposal is well-formed according to our
-        // assumptions about how to store the sent data in our wallet
-        // (num_output_TXOs = num_outlays + change_TXO).
-        if tx_proposal.tx.prefix.outputs.len() - tx_proposal.outlays.len() > 1 {
-            return Err(WalletDbError::UnexpectedNumberOfChangeOutputs);
-        }
+        // // Verify that the TxProposal is well-formed according to our
+        // // assumptions about how to store the sent data in our wallet
+        // // (num_output_TXOs = num_outlays + change_TXO).
+        // if tx_proposal.tx.prefix.outputs.len() - tx_proposal.outlays.len() > 1 {
```
Comment (Shramp) [inline full-service/src/db/transaction_log.rs:416]: delete commented code

File: full-service/src/json_rpc/json_rpc_request.rs [hunk 4]
```diff
@@ -186,12 +190,12 @@ pub enum JsonCommandRequest {
     get_gift_code {
         gift_code_b58: String,
     },
-    get_mc_protocol_transaction {
-        transaction_log_id: String,
-    },
-    get_mc_protocol_txo {
-        txo_id: String,
-    },
+    // get_mc_protocol_transaction {
```
Comment (Shramp) [inline full-service/src/json_rpc/json_rpc_request.rs:193]: delete commented code

File: full-service/src/json_rpc/json_rpc_response.rs [hunk 5]
```diff
@@ -238,12 +237,12 @@ pub enum JsonCommandResponse {
     get_gift_code {
         gift_code: GiftCode,
     },
-    get_mc_protocol_transaction {
-        transaction: JsonTx,
-    },
-    get_mc_protocol_txo {
-        txo: JsonTxOut,
-    },
+    // get_mc_protocol_transaction {
```
Comment (Shramp) [inline full-service/src/json_rpc/json_rpc_response.rs:240]: delete commented code

File: full-service/src/json_rpc/wallet.rs [hunk 6]
```diff
@@ -708,22 +725,22 @@ where
...
-                transaction: json_tx,
-            }
-        }
-        JsonCommandRequest::get_mc_protocol_txo { txo_id } => {
-            let tx_out = service.get_txo_object(&txo_id).map_err(format_error)?;
-            let proto_txo = mc_api::external::TxOut::from(&tx_out);
-            let json_txo = JsonTxOut::from(&proto_txo);
-            JsonCommandResponse::get_mc_protocol_txo { txo: json_txo }
-        }
+        // JsonCommandRequest::get_mc_protocol_transaction { transaction_log_id } => {
```
Comment (Shramp) [inline full-service/src/json_rpc/wallet.rs:728]: delete commented code

**Review by Shramp — CHANGES_REQUESTED** [review]
Looks good! just gotta delete the commented out code and also, is the change to the mobilecoin submodule intentional?

File: docs/accounts/account-secrets/README.md [hunk 7]
```diff
@@ -15,7 +15,7 @@ description: >-
 | `account_id` | string | The unique identifier for the account. |
 | `mnemonic` | string | A BIP39-encoded mnemonic phrase used to generate the account key. |
 | `key_derivation_version` | string \(uint64\) | The version number of the key derivation path used to generate the account key from the mnemonic. |
-| `account_key` | account\_key | The view and spend keys used to transact on the MobileCoin network. Also may contain keys to connect to the Fog ledger scanning service. |
+| `account_key` | account\_key | The view and spend keys used to transact on the MobileCoin network. Also may contain keys to connect to the Fog ledger scanning service. | | `view_account_key` | view\_account\_key | The private view and public spend keys for this account |
```
Comment (sugargoat) [inline docs/accounts/account-secrets/README.md:18]:
```suggestion
| `account_key` | account\_key | The view and spend keys used to transact on the MobileCoin network. Also may contain keys to connect to the Fog ledger scanning service. | 
| `view_account_key` | view\_account\_key | The private view and public spend keys for this account |
```

File: docs/accounts/account-secrets/export_view_only_account_package.md [hunk 8]
```diff
@@ -31,38 +31,12 @@
             "method": "import_view_only_account",
             "params": {
                 "account": {
-                    "object": "view_only_account",
-                    "account_id": "6d95067c5fcc0dd7bbcdd42d49cc3571fe1bb2597a9c397c75b7280eca534208",
```
Comment (sugargoat) [inline docs/accounts/account-secrets/export_view_only_account_package.md:35]: Are these changes relevant to the multi-account support? Can we pull out other changes to another PR?

File: docs/accounts/account/import_view_only_account.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
+---
```
Comment (sugargoat) [inline docs/accounts/account/import_view_only_account.md:1]: Can we move these changes to another PR for improving documentation of view-only accounts?

File: docs/accounts/balance/get_balance_for_account.md [hunk 10]
```diff
@@ -31,18 +31,23 @@ description: Get the current balance for a given account.
...
-      "account_block_height": "152003",
-      "is_synced": false,
-      "unspent_pmob": "110000000000000000",
-      "max_spendable_pmob": "110000000000000000",
-      "pending_pmob": "0",
-      "spent_pmob": "0",
-      "secreted_pmob": "0",
-      "orphaned_pmob": "0"
+    "balance_per_token": {
+      "0": {
```
Comment (sugargoat) [inline docs/accounts/balance/get_balance_for_account.md:35]: It might be nice to prefer the enum label rather than the int value here, especially as the number of tokens grows and the ints have less semantic meaning

File: docs/transactions/transaction-log/README.md [hunk 11]
```diff
@@ -1,10 +1,30 @@
 ---
 description: >-
   A Transaction Log is a record of a MobileCoin transaction that was constructed
-  and sent from this wallet, or that was received at an address belonging to an
```
Comment (sugargoat) [inline docs/transactions/transaction-log/README.md:4]: Are we no longer creating Transaction Logs for received TXOs?

File: full-service/src/json_rpc/e2e_tests/account/account_other.rs [hunk 12]
```diff
@@ -419,8 +419,10 @@ mod e2e_account {
         let txo = &txo_map.get(txos[0].as_str().unwrap()).unwrap();
         let txo_status = txo.get("status").unwrap().as_str().unwrap();
         assert_eq!(txo_status, TxoStatus::Unspent.to_string());
-        let value = txo.get("value_pmob").unwrap().as_str().unwrap();
+        let value = txo.get("value").unwrap();
+        let token_id = txo.get("token_id").unwrap();
         assert_eq!(value, "42000000000000");
+        assert_eq!(token_id, "0");
```
Comment (sugargoat) [inline full-service/src/json_rpc/e2e_tests/account/account_other.rs:425]: I'd love to see an e2e test exercising multi tokens in the same account

File: full-service/src/json_rpc/tx_proposal.rs [hunk 13]
```diff
@@ -2,107 +2,96 @@
...
-                .collect::<Result<Vec<UnspentTxOut>, String>>()?,
-            outlay_list: json_tx_proposal.outlay_list.clone(),
-            tx: json_tx_proposal.tx.clone(),
-            fee: json_tx_proposal.fee.to_string(),
-            outlay_index_to_tx_out_index: outlay_map,
-            outlay_confirmation_numbers: json_tx_proposal.outlay_confirmation_numbers.clone(),
-        })
-    }
+#[derive(Deserialize, Serialize, Default, Debug)]
+pub struct TxProposalJSON {
```
Comment (sugargoat) [inline full-service/src/json_rpc/tx_proposal.rs:28]: I'm not sure about the rename here with the JSON suffix - is the plan to do that for the json types (which are already namespaced by the json_rpc directory?) 

If we do go this way, can we do it in another PR, as it has a pretty large footprint for changes that are not relevant to multi-token

File: full-service/src/json_rpc/json_rpc_request.rs [hunk 14]
```diff
@@ -66,11 +66,12 @@ pub enum JsonCommandRequest {
     },
     build_and_submit_transaction {
         account_id: String,
-        addresses_and_values: Option<Vec<(String, String)>>,
+        addresses_and_values: Option<Vec<(String, String, String)>>,
         recipient_public_address: Option<String>,
-        value_pmob: Option<String>,
+        value: Option<(String, String)>,
```
Comment (sugargoat) [inline full-service/src/json_rpc/json_rpc_request.rs:71]: Can we express Token Type explicitly in the API? I think this proposed API is confusing and error prone, as it's easy to put the wrong string in the wrong index, and it's not deducible what each value represents, or the constraints on that value.

File: full-service/src/json_rpc/json_rpc_request.rs (hunk 4 above)
Comment (sugargoat) [reply full-service/src/json_rpc/json_rpc_request.rs:193]: Why is this being deleted?

File: full-service/src/json_rpc/network_status.rs [hunk 15]
```diff
@@ -9,10 +9,6 @@ use std::{collections::BTreeMap, convert::TryFrom};
 
 #[derive(Deserialize, Serialize, Default, Debug, Clone)]
 pub struct NetworkStatus {
-    /// String representing the object's type. Objects of the same type share
```
Comment (sugargoat) [inline full-service/src/json_rpc/network_status.rs:12]: Can we move these changes to another PR?

File: full-service/src/json_rpc/transaction_log.rs [hunk 16]
```diff
@@ -31,7 +27,7 @@ pub struct TransactionLog {
     pub input_txos: Vec<InputTxo>,
 
     /// A list of the Txos which were outputs from this transaction.
-    pub output_txos: Vec<OutputTxo>,
+    pub payload_txos: Vec<OutputTxo>,
```
Comment (sugargoat) [inline full-service/src/json_rpc/transaction_log.rs:30]: This is potentially a breaking API change. Can we move these changes to another PR and specify the migration plan?

File: full-service/src/json_rpc/transaction_log.rs [hunk 17]
```diff
@@ -113,7 +107,7 @@ impl TransactionLog {
 
 #[derive(Deserialize, Serialize, Default, Debug, Clone)]
 pub struct InputTxo {
-    pub txo_id_hex: String,
+    pub id: String,
```
Comment (sugargoat) [inline full-service/src/json_rpc/transaction_log.rs:110]: This is potentially a breaking API change. Can we move these changes to another PR and specify the migration plan?

File: full-service/src/json_rpc/txo.rs [hunk 18]
```diff
@@ -11,17 +11,15 @@ use serde_derive::{Deserialize, Serialize};
 /// different statuses and types in relation to those accounts.
 #[derive(Deserialize, Serialize, Default, Debug, Clone)]
 pub struct Txo {
-    /// String representing the object's type. Objects of the same type share
-    /// the same value.
-    pub object: String,
```
Comment (sugargoat) [inline full-service/src/json_rpc/txo.rs:16]: Can we move all the object deletion to another PR?

File: full-service/src/json_rpc/txo.rs [hunk 19]
```diff
@@ -11,17 +11,15 @@ use serde_derive::{Deserialize, Serialize};
...
     pub id: String,
 
-    /// Available pico MOB for this account at the current account_block_height.
-    /// If the account is syncing, this value may change.
-    pub value_pmob: String,
+    /// the txo's value
+    pub value: String,
+
+    /// the txo's token id
+    pub token_id: String,
```
Comment (sugargoat) [inline full-service/src/json_rpc/txo.rs:22]: We may want to prefer the token name from enum as opposed to the int

**Review by sugargoat — CHANGES_REQUESTED** [review]
This PR has many changes in it that are not relevant to multi-token. Can we pull out only the multi-token changes for this PR, and move everything else to another PR or separate PRs?

A couple other issues:
1. the API for value on build/submit should be explicit
2. Would we want to prefer enum string (MOB, MOBUSD) vs the enum integer in several places in the API?

File: full-service/src/validator_ledger_sync.rs [hunk 20]
```diff
@@ -145,33 +143,33 @@ impl ValidatorLedgerSyncThread {
...
-        mc_ledger_sync::identify_safe_blocks(ledger_db, &blocks_and_contents, logger)
+        mc_ledger_sync::identify_safe_blocks(ledger_db, &blocks_data, logger)
     }
 
-    fn append_safe_blocks(
-        ledger_db: &mut LedgerDB,
-        blocks_and_contents: &[(Block, BlockContents)],
-        logger: &Logger,
-    ) {
+    fn append_safe_blocks(ledger_db: &mut LedgerDB, block_data: &[BlockData], logger: &Logger) {
```
Comment (sugargoat) [inline full-service/src/validator_ledger_sync.rs:149]: Confirming I understand what's going on in this file - this is a refactor due to upstream refactor around using BlockData directly?

Comment (BrianCorbin) [reply full-service/src/validator_ledger_sync.rs:149]: correct

File: full-service/src/json_rpc/txo.rs (hunk 19 above)
Comment (BrianCorbin) [reply full-service/src/json_rpc/txo.rs:22]: Are all tokens expected to have a name? I haven't looked into that yet, but if so I'm assuming it would be available via a network call or in the MC lib?

File: full-service/src/json_rpc/txo.rs (hunk 18 above)
Comment (BrianCorbin) [reply full-service/src/json_rpc/txo.rs:16]: yes, will clean the PR up and move all unnecessary changes to other pr(s)

File: full-service/src/json_rpc/transaction_log.rs (hunk 17 above)
Comment (BrianCorbin) [reply full-service/src/json_rpc/transaction_log.rs:110]: No direct db migration plan, as this is part of the 2.0 release. As far as the changes for txo_id_hex -> id, I can move that to another PR

File: full-service/src/json_rpc/e2e_tests/transaction/build_submit/large_transaction.rs [hunk 21]
```diff
@@ -66,7 +69,7 @@ mod e2e_transaction {
             "params": {
                 "account_id": account_id,
                 "recipient_public_address": b58_public_address,
-                "value_pmob": "10000000000000000000", // Ten million MOB, which is larger than i64::MAX picomob.
+                "value": ["10000000000000000000", "0"], // Ten million MOB, which is larger than i64::MAX picomob.
```
[the remark this answers is not in the document]
Comment (BrianCorbin) [reply full-service/src/json_rpc/e2e_tests/transaction/build_submit/large_transaction.rs:72]: Full test suite for multi token stuff is incoming :) just wanted to get an initial review

File: docs/transactions/transaction-log/README.md (hunk 11 above)
Comment (BrianCorbin) [reply docs/transactions/transaction-log/README.md:4]: Correct, transaction logs as a concept were reworked in a previous PR and there is not a separation between sent transactions and received (and potentially spent) txos

File: full-service/src/json_rpc/json_rpc_request.rs (hunk 4 above)
Comment (BrianCorbin) [reply full-service/src/json_rpc/json_rpc_request.rs:193]: It isn't, re-enabling it now! had it commented out while working on stuff.

File: full-service/src/json_rpc/tx_proposal.rs (hunk 13 above)
Comment (BrianCorbin) [reply full-service/src/json_rpc/tx_proposal.rs:28]: I agree, we should not have the JSON suffix, and simply import it as an alternal name "*JSON" if necessary.

File: docs/accounts/account/get_account_status.md [hunk 22]
```diff
@@ -42,17 +42,15 @@ description: >-
       "object": "account",
       "recovery_mode": false
     },
-    "balance": {
-      "account_block_height": "152918",
```
[the remark this answers is not in the document]
Comment (BrianCorbin) [reply docs/accounts/account/get_account_status.md:46]: Will address api endpoint discrepancies in the following pr, if that's alright

[the remark this answers is not in the document]
Comment (BrianCorbin) [reply docs/accounts/account/get_account_status.md:46]: will fix documentation changes for this in following PR

File: full-service/src/json_rpc/txo.rs (hunk 19 above)
Comment (sugargoat) [reply full-service/src/json_rpc/txo.rs:22]: Yeah, it's in the submodule, in the protos as `KnownTokenId` https://github.com/mobilecoinfoundation/mobilecoin/blob/master/api/proto/external.proto#L136

And defined in mc-transaction-core in the tokens module as implementing the `Token` trait:
https://github.com/mobilecoinfoundation/mobilecoin/blob/master/transaction/core/src/token.rs#L24

Comment (BrianCorbin) [reply full-service/src/json_rpc/txo.rs:22]: The value of that is still an integer, though. Are you suggesting implementing a conversion within Full Service to map enum values to and from? If so, I would like to do that in a following update as that seems like purely an extension and won't change the API endpoints, will just substitute a stringified Int for a known token name.

File: full-service/src/config.rs [hunk 23]
```diff
@@ -329,7 +329,8 @@ impl LedgerDbConfig {
                     db.append_block(
                         block_data.block(),
                         block_data.contents(),
-                        block_data.signature().cloned(),
+                        block_data.signature(),
+                        block_data.metadata(),
                     )
```
Comment (remoun) [inline full-service/src/config.rs:334]: Optional/Equivalent: Call `append_block_data`

```suggestion
                    db.append_block_data(&block_data)
```

File: full-service/src/json_rpc/json_rpc_request.rs [hunk 24]
```diff
@@ -88,24 +89,27 @@ pub enum JsonCommandRequest {
         txo_id: String,
         output_values: Vec<String>,
         destination_subaddress_index: Option<String>,
-        fee: Option<String>,
+        fee_value: Option<String>,
+        fee_token_id: Option<String>,
```
Comment (remoun) [inline full-service/src/json_rpc/json_rpc_request.rs:93]: Why not `fee: Option<Amount>`?

File: full-service/src/service/gift_code.rs [hunk 25]
```diff
@@ -376,52 +376,62 @@ where
         tombstone_block: Option<u64>,
         max_spendable_value: Option<u64>,
     ) -> Result<(TxProposal, EncodedGiftCode), GiftCodeServiceError> {
-        // First we need to generate a new random bip39 entropy. The way that gift codes
-        // work currently is that the sender creates a middleman account and
-        // sends that account the amount of MOB desired, plus extra to cover the
-        // receivers fee. Then, that account and all of its secrets get encoded
-        // into a b58 string, and when the receiver gets that they can decode it,
+        // // First we need to generate a new random bip39 entropy. The way that
+        // gift codes // work currently is that the sender creates a
```
Comment (remoun) [inline full-service/src/service/gift_code.rs:380]: Nit: Comment needs reformatting

File: full-service/src/service/gift_code.rs [hunk 26]
```diff
@@ -712,289 +722,303 @@ pub fn decode_transfer_payload(
...
-        assert!(service
-            .remove_gift_code(&gift_code_b58)
-            .expect("Could not remove gift code"));
-        let gift_codes = service
-            .list_gift_codes()
-            .expect("Could not list gift codes");
-        assert_eq!(gift_codes.len(), 0);
-    }
-}
+// #[cfg(test)]
```
Comment (remoun) [inline full-service/src/service/gift_code.rs:725]: Nit: Uncomment or delete

File: full-service/src/service/receipt.rs [hunk 27]
```diff
@@ -633,7 +643,14 @@ mod tests {
         let tx_proposal0 = service
             .build_transaction(
                 &alice.id,
-                &vec![(bob_address.to_string(), (24 * MOB).to_string())],
+                &vec![(
+                    bob_address.to_string(),
+                    AmountJSON {
+                        value: (24 * MOB).to_string(),
+                        token_id: Mob::ID.to_string(),
+                    },
```
Comment (remoun) [inline full-service/src/service/receipt.rs:651]: Optional/Equivalent: Add `Amount::new(u64, TokenId)` for type checking

File: full-service/src/validator_ledger_sync.rs [hunk 28]
```diff
@@ -145,33 +143,33 @@ impl ValidatorLedgerSyncThread {
...
         );
 
-        for (block, contents) in blocks_and_contents {
+        for block_data in block_data {
             ledger_db
-                .append_block(block, contents, None)
+                .append_block(
+                    block_data.block(),
+                    block_data.contents(),
+                    block_data.signature(),
```
Comment (remoun) [inline full-service/src/validator_ledger_sync.rs:164]: It turns out we cannot safely append the signature from upstream. I added custom handling in https://github.com/mobilecoinfoundation/mobilecoin/pull/2216, but it might suffice to just pass None here.
```suggestion
                    None,
```

File: full-service/src/json_rpc/json_rpc_request.rs (hunk 24 above)
Comment (BrianCorbin) [reply full-service/src/json_rpc/json_rpc_request.rs:93]: That's intended, if fee_token_id is omitted it will use the default, likewise for fee_value. Neither, one of, and both are valid states.

File: full-service/src/db/transaction_log.rs [hunk 29]
```diff
@@ -405,13 +402,6 @@ impl TransactionLogModel for TransactionLog {
         // Verify that the account exists.
         Account::get(&AccountID(account_id_hex.to_string()), conn)?;
 
-        // Verify that the TxProposal is well-formed according to our
-        // assumptions about how to store the sent data in our wallet
-        // (num_output_TXOs = num_outlays + change_TXO).
-        if tx_proposal.tx.prefix.outputs.len() - tx_proposal.outlays.len() > 1 {
```
Comment (sugargoat) [inline full-service/src/db/transaction_log.rs:411]: What's the reason this invariant no longer holds?

File: full-service/src/db/txo.rs [hunk 30]
```diff
@@ -354,85 +351,45 @@ impl TxoModel for Txo {
...
-            subaddress_index,
-            key_image: None, // Only the recipient can calculate the KeyImage
+            value: output_txo.value as i64,
+            token_id: *output_txo.token_id as i64,
+            target_key: &mc_util_serial::encode(&output_txo.tx_out.target_key),
+            public_key: &mc_util_serial::encode(&output_txo.tx_out.public_key),
+            e_fog_hint: &mc_util_serial::encode(&output_txo.tx_out.e_fog_hint),
+            txo: &mc_util_serial::encode(&output_txo.tx_out),
+            subaddress_index: None,
+            key_image: None,
```
Comment (sugargoat) [inline full-service/src/db/txo.rs:375]: I think the comment is nice to explain why a KeyImage is None for this type of txo

```suggestion
            key_image: None, // Only the recipient can calculate the KeyImage
```

File: full-service/src/db/txo.rs [hunk 31]
```diff
@@ -354,85 +351,45 @@ impl TxoModel for Txo {
         Ok(txo_id.to_string())
     }
 
-    fn create_minted(
-        output: &TxOut,
-        tx_proposal: &TxProposal,
```
Comment (sugargoat) [inline full-service/src/db/txo.rs:359]: Does this still work with the offline flow, where we are copying TxProposals from the offline machine? Or is the assumption that the online machine is never storing data about which TXOs they minted, and that path is now removed?

File: full-service/src/json_rpc/e2e_tests/transaction/build_submit/build_and_submit.rs [hunk 32]
```diff
@@ -151,6 +160,144 @@ mod e2e_transaction {
...
+        assert!(err.is_some());
+
+        let body = json!({
+            "jsonrpc": "2.0",
+            "id": 1,
+            "method": "build_and_submit_transaction",
+            "params": {
+                "account_id": account_id,
+                "recipient_public_address": b58_public_address,
+                "amount": { "value": "500000000000", "token_id": "1" }
```
Comment (sugargoat) [inline full-service/src/json_rpc/e2e_tests/transaction/build_submit/build_and_submit.rs:271]: 🙌

File: full-service/src/json_rpc/e2e_tests/transaction/build_submit/build_and_submit.rs [hunk 33]
```diff
@@ -174,5 +369,17 @@ mod e2e_transaction {
...
+        let unspent = balance_1["unspent"].as_str().unwrap();
+        let pending = balance_1["pending"].as_str().unwrap();
+        let spent = balance_1["spent"].as_str().unwrap();
+        let secreted = balance_1["secreted"].as_str().unwrap();
+        let orphaned = balance_1["orphaned"].as_str().unwrap();
+        assert_eq!(unspent, "999999998976".to_string());
+        assert_eq!(pending, "0");
+        assert_eq!(spent, "1000000000000");
+        assert_eq!(secreted, "0");
+        assert_eq!(orphaned, "0");
```
Comment (sugargoat) [inline full-service/src/json_rpc/e2e_tests/transaction/build_submit/build_and_submit.rs:383]: Awesome, great test, ty!

File: full-service/src/json_rpc/json_rpc_request.rs [hunk 34]
```diff
@@ -66,11 +66,12 @@ pub enum JsonCommandRequest {
     },
     build_and_submit_transaction {
         account_id: String,
-        addresses_and_values: Option<Vec<(String, String)>>,
+        addresses_and_amounts: Option<Vec<(String, Amount)>>,
```
Comment (sugargoat) [inline full-service/src/json_rpc/json_rpc_request.rs:69]: To make this backward compatible, we can leave the old value, and if it is None, use the new value, and if it is populated, use it.

```suggestion
        # Multi-token Support
        addresses_and_amounts: Option<Vec<(String, Amount)>>,
        # MOB-only
        addresses_and_values: Option<Vec<(String, String)>>,
```

File: full-service/src/json_rpc/json_rpc_request.rs [hunk 35]
```diff
@@ -66,11 +66,12 @@ pub enum JsonCommandRequest {
...
         recipient_public_address: Option<String>,
-        value_pmob: Option<String>,
+        amount: Option<Amount>,
```
Comment (sugargoat) [inline full-service/src/json_rpc/json_rpc_request.rs:71]: Similarly here:

```suggestion
        // Multi-token Support
        amount: Option<Amount>,
        // MOB-Only
        value_pmob: Option<String>
```

File: full-service/src/json_rpc/json_rpc_request.rs [hunk 36]
```diff
@@ -66,11 +66,12 @@ pub enum JsonCommandRequest {
...
         input_txo_ids: Option<Vec<String>>,
-        fee: Option<String>,
+        fee_value: Option<String>,
```
Comment (sugargoat) [inline full-service/src/json_rpc/json_rpc_request.rs:73]: For this one, I would suggest keeping to one param, and having a backward compatible param

```suggestion
        // Multi-token Support
        fee_amount: Option<Amount>,
        // MOB-Only
        fee: Option<String>,
```

File: full-service/src/json_rpc/json_rpc_request.rs [hunk 37]
```diff
@@ -88,24 +89,27 @@ pub enum JsonCommandRequest {
         txo_id: String,
         output_values: Vec<String>,
         destination_subaddress_index: Option<String>,
-        fee: Option<String>,
+        fee_value: Option<String>,
```
Comment (sugargoat) [inline full-service/src/json_rpc/json_rpc_request.rs:92]:
```suggestion
        // Multi-token Support
        fee_amount: Option<Amount>,
        // MOB-Only
        fee: Option<String>,
```

File: full-service/src/json_rpc/json_rpc_request.rs [hunk 38]
```diff
@@ -88,24 +89,27 @@ pub enum JsonCommandRequest {
...
+        fee_token_id: Option<String>,
         tombstone_block: Option<String>,
     },
     build_transaction {
         account_id: String,
-        addresses_and_values: Option<Vec<(String, String)>>,
+        addresses_and_amounts: Option<Vec<(String, Amount)>>,
```
Comment (sugargoat) [inline full-service/src/json_rpc/json_rpc_request.rs:98]:
```suggestion
        // Multi-token Support
        addresses_and_amounts: Option<Vec<(String, Amount)>>,
        // MOB-only
        addresses_and_values: Option<Vec<(String, String)>>,
```

File: full-service/src/json_rpc/json_rpc_request.rs [hunk 39]
```diff
@@ -88,24 +89,27 @@ pub enum JsonCommandRequest {
...
         recipient_public_address: Option<String>,
-        value_pmob: Option<String>,
+        amount: Option<Amount>,
```
Comment (sugargoat) [inline full-service/src/json_rpc/json_rpc_request.rs:100]:
```suggestion
        // Multi-Token Support
        amount: Option<Amount>,
        // MOB-only
        value_pmob: Option<String>,
```

File: full-service/src/json_rpc/json_rpc_request.rs [hunk 40]
```diff
@@ -88,24 +89,27 @@ pub enum JsonCommandRequest {
...
         input_txo_ids: Option<Vec<String>>,
-        fee: Option<String>,
+        fee_value: Option<String>,
+        fee_token_id: Option<String>,
```
Comment (sugargoat) [inline full-service/src/json_rpc/json_rpc_request.rs:103]:
```suggestion
        // Multi-Token Support
        fee_amount: Option<Amount>,
        // MOB-Only
        fee: Option<String>,
```

File: full-service/src/json_rpc/json_rpc_request.rs [hunk 41]
```diff
@@ -88,24 +89,27 @@ pub enum JsonCommandRequest {
...
         max_spendable_value: Option<String>,
     },
     build_unsigned_transaction {
         account_id: String,
         recipient_public_address: Option<String>,
-        value_pmob: Option<String>,
-        fee: Option<String>,
+        amount: Option<Amount>,
+        fee_value: Option<String>,
+        fee_token_id: Option<String>,
```
Comment (sugargoat) [inline full-service/src/json_rpc/json_rpc_request.rs:112]:
```suggestion
        // Multi-Token Support
        amount: Option<Amount>,
        // MOB-Only
        value_pmob: Option<String>,
        // Multi-Token Support
        fee_amount: Option<Amount>,
        // MOB-Only
        fee: Option<String>,
```

File: full-service/src/service/models/tx_proposal.rs [hunk 42]
```diff
@@ -0,0 +1,120 @@
+use std::convert::TryInto;
```
Comment (sugargoat) [inline full-service/src/service/models/tx_proposal.rs:1]: Ah, I see - this is where tx_proposal lives now in the DB

File: full-service/src/db/txo.rs (hunk 31 above)
Comment (BrianCorbin) [reply full-service/src/db/txo.rs:359]: Information about minted txos is still available, but in the TransactionOutputs table. The minted account field in the txos table was a duplicate data entry

File: full-service/src/db/txo.rs (hunk 30 above)
Comment (BrianCorbin) [reply full-service/src/db/txo.rs:375]: Fair point, this may have been an unintentional removal during my mad frenzy

File: full-service/src/json_rpc/json_rpc_request.rs (hunk 41 above)
Comment (BrianCorbin) [reply full-service/src/json_rpc/json_rpc_request.rs:112]: fee_token_id and fee_value are not both required, as if only 1 is provided then the API will figure out the default for the other.

If fee_token_id is set to "0" and fee_value is omitted, it will use the default minimum if available.

File: full-service/src/db/transaction_log.rs (hunk 29 above)
Comment (BrianCorbin) [reply full-service/src/db/transaction_log.rs:411]: We can have multiple change txos once mixed token transactions are enabled, and I also cannot think of a scenario where this is even possible when there is a single token type in the transaction, so it seems unnecessary to check for.

**Review by Shramp — APPROVED** [review]
lgtm
